File: SMART/app/src/main/java/com/cs565project/smart/fragments/SetRestrictionFragment.java
package com.cs565project.smart.fragments;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;

import mobi.upod.timedurationpicker.TimeDurationPicker;
import mobi.upod.timedurationpicker.TimeDurationPickerDialogFragment;

/**
 * Dialog fragment to show a time duration picker.
 */
public class SetRestrictionFragment extends TimeDurationPickerDialogFragment {

    // the fragment initialization parameter keys.
    private static final String ARG_APP_NAME = "app_name";
    private static final String ARG_PACKAGE_NAME = "package_name";
    private static final String ARG_DEFAULT_VAL = "default_val";

    private String myAppName, myPackageName;
    private long myDefaultValue;

    private OnDurationSelectedListener mListener;

    public SetRestrictionFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     */
    public static SetRestrictionFragment newInstance(String appName, String packageName, long defaultValue) {
        SetRestrictionFragment fragment = new SetRestrictionFragment();
        Bundle args = new Bundle();
        args.putString(ARG_APP_NAME, appName);
        args.putLong(ARG_DEFAULT_VAL, defaultValue);
        args.putString(ARG_PACKAGE_NAME, packageName);
        fragment.setArguments(args);
        return fragment;
    }

    public SetRestrictionFragment setListener(OnDurationSelectedListener listener) {
        mListener = listener;
        return this;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            myAppName = getArguments().getString(ARG_APP_NAME);
            myDefaultValue = getArguments().getLong(ARG_DEFAULT_VAL);
            myPackageName = getArguments().getString(ARG_PACKAGE_NAME);
        }
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        dialog.setTitle(String.format("Set daily allowed duration for %s", myAppName));
        return dialog;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onDurationSet(TimeDurationPicker view, long duration) {
        if (mListener != null) {
            mListener.onDurationConfirmed(myPackageName, duration);
        }
    }

    @Override
    public void onCancel(DialogInterface dialog) {
        super.onCancel(dialog);
        if (mListener != null) {
            mListener.onCancel();
        }
    }

    @Override
    protected long getInitialDuration() {
        return myDefaultValue;
    }

    @Override
    protected int setTimeUnits() {
        return TimeDurationPicker.HH_MM;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     */
    public interface OnDurationSelectedListener {
        void onDurationConfirmed(String packageName, long duration);

        void onCancel();
    }
}

File: SMART/app/src/main/java/com/cs565project/smart/fragments/SetRestrictionFragment.kt
package com.cs565project.smart.fragments

import android.content.DialogInterface
import android.os.Bundle
import android.support.v7.app.AppCompatDialog
import android.support.v7.app.AppCompatDialogFragment

import com.cs565project.smart.R

import mobi.upod.timedurationpicker.TimeDurationPicker
import mobi.upod.timedurationpicker.TimeDurationPickerDialog

/**
 * Dialog fragment to show a time duration picker.
 */
class SetRestrictionFragment : AppCompatDialogFragment(), TimeDurationPickerDialog.OnDurationSetListener {

    private var myAppName: String? = null
    private var myPackageName: String? = null
    protected var initialDuration: Long = 0
        private set

    private var mListener: OnDurationSelectedListener? = null

    fun setListener(listener: OnDurationSelectedListener): SetRestrictionFragment {
        mListener = listener
        return this
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            myAppName = arguments!!.getString(ARG_APP_NAME)
            initialDuration = arguments!!.getLong(ARG_DEFAULT_VAL)
            myPackageName = arguments!!.getString(ARG_PACKAGE_NAME)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): AppCompatDialog {
        val dialog = TimeDurationPickerDialog(activity!!, this, initialDuration, setTimeUnits())
        dialog.setTitle(String.format(getString(R.string.set_restriction_title), myAppName))
        return dialog
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun onDurationSet(view: TimeDurationPicker, duration: Long) {
        if (mListener != null) {
            mListener!!.onDurationConfirmed(myPackageName, duration)
        }
    }

    override fun onCancel(dialog: DialogInterface?) {
        super.onCancel(dialog)
        if (mListener != null) {
            mListener!!.onCancel()
        }
    }

    protected fun setTimeUnits(): Int {
        return TimeDurationPicker.HH_MM
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     */
    interface OnDurationSelectedListener {
        fun onDurationConfirmed(packageName: String?, duration: Long)

        fun onCancel()
    }

    companion object {

        // the fragment initialization parameter keys.
        private const val ARG_APP_NAME = "app_name"
        private const val ARG_PACKAGE_NAME = "package_name"
        private const val ARG_DEFAULT_VAL = "default_val"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         */
        fun newInstance(appName: String, packageName: String, defaultValue: Long): SetRestrictionFragment {
            val fragment = SetRestrictionFragment()
            val args = Bundle()
            args.putString(ARG_APP_NAME, appName)
            args.putLong(ARG_DEFAULT_VAL, defaultValue)
            args.putString(ARG_PACKAGE_NAME, packageName)
            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
